[ONNX] 'Transform' as base class for passes

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96781
* #95651
* #96935
* __->__ #95935

Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.
New base class `Pass` will be added when `analysis` is introduced.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">

BowenBao committed:
[ONNX] 'Pass' and 'Transform' as base class for passes

[ghstack-poisoned]

BowenBao commented:
CI failure unrelated, regression in atenlib `native_layer_norm`

wschin commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -0,0 +1,93 @@
+    def compare(self, a: Sequence[str], b: Sequence[str]) -> Generator[str, None, None]:
+        cruncher = difflib.SequenceMatcher(None, a, b, autojunk=False)
+        for tag, alo, ahi, blo, bhi in cruncher.get_opcodes():
+            if tag == "replace":
```
Where does those magic keywords `replace` , `delete`, `insert`, etc come from? Can you add some example links to `SequenceMatcher` so that people knows what does it really do? Just like what I did for ONNX external data.

BowenBao replied:
these are gone now

wschin commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -0,0 +1,93 @@
+    `Reference: Automatic junk heuristic <https://docs.python.org/3/library/difflib.html>`_
+    """
+
+    def compare(self, a: Sequence[str], b: Sequence[str]) -> Generator[str, None, None]:
```
I can see all the type annotations, but I still don't know what does this function do. Can you provide some example inputs and example outputs?

BowenBao replied:
docstring is up there which mentioned the difference between this and the official Differ class, and why we need it.

Please let me know if you see there is a better way, I don't like having to "copy" all the code here neither.

BowenBao replied:
For now I'm adding a docstring for compare to mention this again. I'm open to other suggestions.

BowenBao replied:
I tried another approach with context manager, please take a look. And I didn't add docstring for compare, that way it inherits parent's docstring. Added comment for context manager.

wschin commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -0,0 +1,93 @@
+            yield from g
+
+
+class Pass(abc.ABC):
```
What is this for? Can we remove it and  do
```
class Transform(abc.ABC))
```

BowenBao replied:
For future analysis, but you are right for the context of this PR it makes sense to just use Transform as base.

BowenBao committed:
Update how 'compare' is patched on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Support more passes & more docstrings on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
bump onnxscript on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Pass' and 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Improve perf of graph diff on "[ONNX] 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.
New base class `Pass` will be added when `analysis` is introduced.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.
New base class `Pass` will be added when `analysis` is introduced.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.
New base class `Pass` will be added when `analysis` is introduced.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/symbolic_exporter.py:
```diff
@@ -266,12 +195,16 @@ def export_without_parameters_and_buffers(
+    replaced_attrs = replace_get_attr_with_placeholder_pass.replaced_attrs
     # Move all newly created placeholder nodes to the front of the graph.
-    _move_placeholder_to_front(graph_module)
+    graph_module = passes.MovePlaceholderToFront(graph_module).run()
```
What happens if run was called twice?

wschin replied:
No thing happens in 2nd run. It will be good to add a test.

BowenBao replied:
Can't guarantee for all passes though. I kinda want to make it functional, but it seems there are cases where the state is useful. For now I don't worry too much about this, but let me know if you have particular concern.

BowenBao replied:
And at least diagnostics catches it running twice, and the callstack that leads to that.

BowenBao replied:
Could add a guard in the class against multi run too if we really hate it.

justinchuby replied:
I am not really against multiple runs. For iterative passes we might actually want it? Just trying to clarify and understand the behavior

justinchuby reviewed (approved):
beautiful 🏡

justinchuby commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -0,0 +1,204 @@
+    def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
+        ...
+
+    @diagnostics.diagnose_call(
```
Thoughts on creating a hook on `run` to allow injecting `diagnostics.diagnose_call`?

BowenBao replied:
Could you share a snippet maybe, what is the main benefit?

justinchuby replied:
Just a thought. May not be a good idea now that I think about it

BowenBao committed:
Update on "[ONNX] 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.
New base class `Pass` will be added when `analysis` is introduced.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

wschin commented on torch/onnx/_internal/fx/exporter.py:
```diff
@@ -36,14 +36,14 @@ def _export(
     # Make sure the feed-in "module" is stateless.
     # Ensure placeholder targets match the original module's signature since
```
Remove the comments moved into `Decompose`.

BowenBao committed:
Update on "[ONNX] 'Transform' as base class for passes"


Base class `Transform` provides basic diagnostics functionality. Diagnostics
are automatically recorded for inherited passes.
New base class `Pass` will be added when `analysis` is introduced.

Example diagnostics for `test_mnist`:

Decompose:
<img src="https://user-images.githubusercontent.com/9376104/222615465-689e76eb-6b30-4670-aed5-a0d419583bfe.png" width="80%" height="80%">

Shape inference:
<img src="https://user-images.githubusercontent.com/9376104/222615527-0484e504-f9d5-4f5c-b018-3e45ef15c138.png" width="80%" height="80%">

Moving placeholders:
<img src="https://user-images.githubusercontent.com/9376104/222852379-36caf263-6965-4e5d-9dce-f63075a3812f.png" width="80%" height="80%">



[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g

titaiwangms commented:
Oh no, this one is merged...
